internal/users/manager_bwrap_test.go

```diff
@@ -0,0 +1,419 @@
+				setHome(t, m, "user1", home)
+			}
+
+			newGID := 54321
+			if tc.gidTooLarge {
+				newGID = math.MaxInt32 + 1
+			}
+
+			if tc.gidAlreadySet {
+				setGID(t, m, groupname, newGID)
+			}
+			if tc.gidAlreadyInUseByAuthdGroup {
+				setGID(t, m, "group3", newGID)
+			}
+			if tc.gidAlreadyInUseAsUIDofAuthdUser {
+				setUID(t, m, "user2", newGID)
+			}
+			if tc.gidAlreadyInUseBySystemGroup {
+				addGroupToSystem(t, newGID)
+			}
+			if tc.gidAlreadyInUseAsUIDofSystemUser {
+				addUserToSystem(t, newGID)
+			}
+			if tc.gidIsNotPrimaryGroupOfAnyUser {
+				// Change the primary group of "user1" to another group
+				setPrimaryGroup(t, m, "user1", 22222)
+			}
+			if tc.gidIsPrimaryGroupOfMultipleUsers {
+				// Change the primary group of "user2" to the group we want to change
+				setPrimaryGroup(t, m, "user2", 11111)
+			}
+
+			//nolint:gosec // G115 we set the GID above to values that are valid uint32
+			warnings, err := m.SetGroupID(groupname, uint32(newGID))
+			log.Infof(context.Background(), "SetGroupID error: %v", err)
+			log.Infof(context.Background(), "SetGroupID warnings: %v", warnings)
+
+			if tc.wantErrType != nil {
+				require.ErrorIs(t, err, tc.wantErrType, "SetGroupID should return expected error")
+				return
+			}
+			if tc.wantErr {
+				require.Error(t, err, "SetGroupID should return an error but didn't")
+				return
+			}
+			require.NoError(t, err, "SetGroupID should not return an error")
+			require.Len(t, warnings, tc.wantWarnings, "Unexpected number of warnings")
+
+			yamlData, err := db.Z_ForTests_DumpNormalizedYAML(m.DB())
+			require.NoError(t, err)
+			golden.CheckOrUpdate(t, yamlData, golden.WithPath("db"))
+
+			if len(warnings) == 0 {
+				return
+			}
+
+			// To make the tests deterministic, we replace the temporary home directory path with a placeholder
+			for i, w := range warnings {
+				if regexp.MustCompile(`Could not get owner of home directory "([^"]+)"`).MatchString(w) {
+					warnings[i] = `Could not get owner of home directory "{{HOME}}"`
+				}
+				if regexp.MustCompile(`Not changing ownership of home directory "([^"]+)", because it is not owned by GID \d+ \(current owner: \d+\)`).MatchString(w) {
+					warnings[i] = `Not changing ownership of home directory "{{HOME}}", because it is not owned by GID {{GID}} (current owner: {{CURR_GID}})`
+				}
+			}
+
+			golden.CheckOrUpdateYAML(t, warnings, golden.WithPath("warnings"))
+		})
+	}
+}
+
+// createTemporaryHome creates a temporary home directory for the given user.
+func createTemporaryHome(t *testing.T, uid, gid int, inaccessible, cannotBeChanged bool) string {
+	t.Helper()
+
+	// We use a deterministic path (/tmp/home) here because the home directory
+	// is stored in the database which we dump to a golden file, so we would
+	// have to replace the path in the golden file to make it deterministic.
+	// It's simpler to just use a deterministic path here.
```

Comment:
This looks like another reason to do something like https://github.com/ubuntu/authd/pull/979/commits/f7c0622b5c5dc1f064c31b4ff7c2b817de6d3095#diff-e8a1ba2baeae8e0a9bceba1f716778bab92beb602eda806dffeddbbd54f2924b as IIRC was done to make possible to make the users home deterministic, not sure if the same applies here though :)

Reply:
I find the solution to use a deterministic path below `/tmp` simple and quite elegant - given that the test is executed with bwrap, so the deterministic path doesn't cause conflicts with other tests.

Reply:
> BTW... Ideally all the integration tests should maybe just go through bubblewrap, so that we can simulate a real scenario better without env variables or preloading hacks.

yes, I also suspect we could simplify things a lot and make tests closer to real scenarios by using bwrap more